File: src/services/signinService/signinService.go
package signinService

import (
	"context"
	"i9rfs/src/appTypes"
	user "i9rfs/src/models/userModel"
	"i9rfs/src/services/securityServices"
	"os"
	"time"

	"github.com/gofiber/fiber/v2"
)

func Signin(ctx context.Context, emailOrUsername, password string) (any, string, error) {
	theUser, err := user.AuthFind(ctx, emailOrUsername)
	if err != nil {
		return nil, "", err
	}

	if theUser == nil {
		return nil, "", fiber.NewError(fiber.StatusNotFound, "Incorrect email or password")
	}

	hashedPassword := theUser["password"].(string)

	yes, err := securityServices.PasswordMatchesHash(hashedPassword, password)
	if err != nil {
		return nil, "", err
	}

	if !yes {
		return nil, "", fiber.NewError(fiber.StatusNotFound, "Incorrect email or password")
	}

	authJwt, err := securityServices.JwtSign(appTypes.ClientUser{
		Username: theUser["username"].(string),
	}, os.Getenv("AUTH_JWT_SECRET"), time.Now().UTC().Add(10*24*time.Hour))

	if err != nil {
		return nil, "", err
	}

	delete(theUser, "password")

	respData := map[string]any{
		"msg":  "Signin success!",
		"user": theUser,
	}

	return respData, authJwt, nil
}

File: src/services/signinService.ts
import { StatusCodes } from "http-status-codes"
import * as user from "../models/userModel.ts"
import * as securityServices from "./util/securityServices.ts"
import type { ClientUserT } from "../appTypes.ts"

interface SigninResp {
  msg: string
  user: ClientUserT
}

export async function Signin(emailOrUsername: string, inputPassword: string) {
  const theUser = await user.AuthFind(emailOrUsername)

  if (!theUser) {
    throw {
      name: "AppError",
      code: StatusCodes.NOT_FOUND,
      message: "Incorrect email or password",
    }
  }

  const yes = await securityServices.PasswordMatchesHash(
    theUser.password,
    inputPassword
  )

  if (!yes) {
    throw {
      name: "AppError",
      code: StatusCodes.NOT_FOUND,
      message: "Incorrect email or password",
    }
  }

  const authJwt = securityServices.JwtSign(
    {
      username: theUser.username,
    },
    process.env.AUTH_JWT_SECRET || "",
    10 * 24 * 60 * 60 * 1000
  ) // 10 days

  const respData: SigninResp = {
    msg: "Signin success!",
    user: { username: theUser.username },
  }

  return { respData, authJwt }
}
